src/imitation/algorithms/preference_comparisons.py

```diff
@@ -981,16 +1187,26 @@ def __init__(
             weight_decay,
             custom_logger,
         )
+        self.rng = np.random.default_rng(seed=seed)
 
     def _training_inner_loop(
         self,
         fragment_pairs: Sequence[TrajectoryPair],
         preferences: np.ndarray,
     ) -> th.Tensor:
+        assert len(fragment_pairs) == preferences.shape[0]
         losses = []
         metrics = []
-        for member in self._model.members:
-            output = self.loss.forward(member, fragment_pairs, preferences)
+        for member_idx in range(self._model.num_members):
```

Comment:
Is bagging the batch what we want to do here? Or should we be sampling independent batches from the dataset for each ensemble member?

Reply:
Independent batches does seem better. The standard bagging approach would be to subsample the entire dataset -- so if the batch size `len(fragment_pairs)` is entire dataset we recover that. But with smaller batch sizes the bagged batches will end up quite correlated with one another.

In the interests of wrapping this up soon, I'd be fine with merging the PR with the current bagging approach and just opening an issue to track it, then improve it later.

Reply:
Yes, I'll add the correct implementation of bagging over the entire dataset through another PR. 

The current version is sort of an intermediate option of no bagging vs. bagging over the entire dataset. It shouldn't perform worse than no bagging IMO.

Reply:
Agreed. Also, I'd be happy to write the improved bagging code latter today.